lib/ansible/plugins/httpapi/fortios.py

```diff
@@ -0,0 +1,137 @@
+# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+# IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+# USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+DOCUMENTATION = """
+---
+author:
+    - Miguel Angel Muñoz (@magonzalez)
+httpapi : fortios
+short_description: HttpApi Plugin for Fortinet FortiOS Appliance or VM
+description:
+  - This HttpApi plugin provides methods to connect to Fortinet FortiOS Appliance or VM via REST API
+version_added: "2.9"
+
+"""
+
+from ansible.plugins.httpapi import HttpApiBase
+from ansible.module_utils.basic import to_text
+import urllib
+import json
+import re
+
+
+class HttpApi(HttpApiBase):
+    def __init__(self, connection):
+        super(HttpApi, self).__init__(connection)
+
+        self._connection = connection
+        self._become = False
+        self._become_pass = ''
+        self._ccsrftoken = ''
+
+    def set_become(self, become_context):
+        """
+        Elevation is not required on Fortinet devices - Skipped
+        :param become_context: Unused input.
+        :return: None
+        """
+        return None
+
+    def login(self, username, password):
+        """Call a defined login endpoint to receive an authentication token."""
+
+        data = "username=" + urllib.parse.quote(username) + "&secretkey=" + urllib.parse.quote(password) + "&ajax=1"
+        dummy, result_data = self.send_request(url='/logincheck', data=data, method='POST')
+        if result_data[0] != '1':
+            raise Exception('Wrong credentials. Please check')
+
+    def logout(self):
+        """ Call to implement session logout."""
+
+        self.send_request(url='/logout', method="POST")
+
+    def update_auth(self, response, response_text):
+        """
+        Get cookies and obtain value for csrftoken that will be used on next requests
+        :param response: Response given by the server.
+        :param response_text Unused_input.
+        :return: Dictionary containing cookies
+        """
+
+        cookies = {}
+
+        for attr, val in response.getheaders():
+            if attr == 'Set-Cookie' and 'APSCOOKIE_' in val:
+                cookies['Cookie'] = val
+
+            elif attr == 'Set-Cookie' and 'ccsrftoken=' in val:
+                csrftoken_search = re.search('\"(.*)\"', val)
+                if csrftoken_search:
+                    self._ccsrftoken = csrftoken_search.group(1)
```

Comment:
And as the return is a dictionary of headers, instead of tracking `self._ccsrftoken` here, you should be able to set your return dictionary to be `{'x-csrftoken': csrftoken_search.group(1)}`, and everything should work as expected.

Reply:
Thanks Qalthos, I have just tried your proposal but I am afraid I have to store the csrctoken anyway. The reason is that our fortios device sends the token on the reply to the first login request only. That means that the next request after the login works fine, as it gets the headers correctly with the csrftoken, but next requests do not get the csrftoken anymore and fortios requires the csrftoken on every request.

Reply:
If you're only going to get `APSCOOKIE_` _or_ `ccsrftoken`, then you can just return `None` if you don't find anything and the existing token will be reused.

If you are expecting to have both, then I would just dedent the next line to be outside the for loop, so that the token is always added to the dictionary on every run. Then you should be able to at least remove the manual headers building in `send_request`.

Reply:
Ok, I'd go for the latter option as it seems safer for our fortios. However, how can I access the headers in 'send_request' method? I need to know if there is a csrftoken set bcs if not I must raise an exception to instruct the user to log in first.

Reply:
Hi again @Qalthos, I have been doing some more additional tests and I think I have no means to access the headers in 'send_request' method (which I need to know if there is a csrf_token from a previous login before). Therefore I will have to leave the code it as it is now. I would mark this as resolved, please let me know if it is not ok or if I missed something and you want me to perform more changes.

Reply:
Hi Qalthos,
 thanks for your answer. If the system guarantees that the first call is a login then I think you are right and the exception will never raise. I can remove that part, thanks for the clarification. I just added it as an extra 'safety' measure.

Regarding the csrf token lost, our device sends the csrf token and the cookie in the first login, but in the later request it may send only the cookie, or nothing. There is no rule about that. However whenever we send a PUT or POST we must add both the token and the cookie always. And here is where I found that if I don't store the csrf token it will be lost in later requests. I have run some tests, and the ansible system do not keep the csrftoken after the first request (even if I return "None" in update_auth). Therefore, I see no other way to make it work than storing the csrf token in the class. Please let me know if I am doing something wrong. This is the code I am testing according to your comments but it does not work bcs later requests do not keep the csrf token:
```
    def update_auth(self, response, response_text):

        headers = {}

        for attr, val in response.getheaders():
            if attr == 'Set-Cookie' and 'APSCOOKIE_' in val:
                headers['Cookie'] = val

            elif attr == 'Set-Cookie' and 'ccsrftoken=' in val:
                csrftoken_search = re.search('\"(.*)\"', val)
                if csrftoken_search:
                    headers['x-csrftoken'] = self._ccsrftoken

        if len (headers) == 2:
            return headers
        else:
            return None

    def send_request(self, **message_kwargs):

        url = message_kwargs.get('url', '/')
        data = message_kwargs.get('data', '')
        method = message_kwargs.get('method', 'GET')

        try:
            response, response_data = self.connection.send(url, data, #headers=headers,
                                                           method=method)

            return response.status, to_text(response_data.getvalue())
        except Exception as err:
            raise Exception(err)
```
Please let me know if I am doing something wrong or if I misunderstood something.
Thanks!